### PR TITLE
Fix PlainTableReader not to crash sst_dump

### PR DESCRIPTION
**Pull request by pdillinger**
Summary: Plain table SSTs could crash sst_dump because of a bug in
PlainTableReader that can leave table_properties_ as null. Even if it
was intended not to keep the table properties in some cases, they were
leaked on the offending code path.

Steps to reproduce:

    $ db_bench --benchmarks=fillrandom --num=2000000 --use_plain_table --prefix-size=12
    $ sst_dump --file=0000xx.sst --show_properties
    from [] to []
    Process /dev/shm/dbbench/000014.sst
    Sst file format: plain table
    Raw user collected properties
    ------------------------------
    Segmentation fault (core dumped)

Also added missing unit testing of plain table full_scan_mode, and
an assertion in NewIterator to check for regression.

Test Plan: new unit test, manual, make check

### CODE REVIEW

File: db/plain_table_db_test.cc [hunk 1]
```diff
@@ -454,20 +457,36 @@ TEST_P(PlainTableDBTest, Flush) {
...
+              ASSERT_EQ("0", (tp->user_collected_properties)
+                                 .at("plain_table_sub_index_size"));
+            } else {
+              ASSERT_EQ("0", (tp->user_collected_properties)
+                                 .at("plain_table_hash_table_size"));
+              ASSERT_EQ("0", (tp->user_collected_properties)
+                                 .at("plain_table_sub_index_size"));
+            }
+            ASSERT_EQ("v3", Get("1000000000000foo"));
+            ASSERT_EQ("v2", Get("0000000000000bar"));
```
Comment (siying) [inline db/plain_table_db_test.cc]: Are these changes related to the fix?

Comment (pdillinger) [reply db/plain_table_db_test.cc]: There's new code under `if (full_scan_mode) {`. Otherwise, it's just whitespace changes.

File: table/plain/plain_table_reader.cc [hunk 2]
```diff
@@ -183,6 +183,8 @@ Status PlainTableReader::Open(
     // can be used.
     new_reader->full_scan_mode_ = true;
   }
+  // PopulateIndex can add to the props, so don't store them until now
+  new_reader->table_properties_.reset(props);
```
Comment (siying) [inline table/plain/plain_table_reader.cc]: Should we move the same line inside PopulateIndex()? It feels confusing to me to set it twice but happens to be the same one to be correct.

**Comment by pdillinger** [comment]
AppVeyor timeout during tests

File: table/plain/plain_table_reader.cc (hunk 2 above)
Comment (pdillinger) [reply table/plain/plain_table_reader.cc]: Old state: only set in PlainTableReader::PopulateIndex, which is not always called. New state: only/always set in PlainTableReader::Open (and *Test*PlainTableReader ctor since it does not use PlainTableReader::Open).
